ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,275 @@
+
+
+def _traced_parse(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    # If graphql.parse() is called outside graphql.graphql(), graphql.parse will
+    # be a top level span. Thereforce we must explicitly set the service name.
+    with pin.tracer.trace(
+        name="graphql.parse",
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ):
+        return func(*args, **kwargs)
+
+
+def _traced_validate(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    # If graphql.parse() is called outside graphql.graphql(), graphql.parse will
+    # be a top level span. Thereforce we must explicitly set the service name.
+    with pin.tracer.trace(
+        name="graphql.validate",
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ) as span:
+        errors = func(*args, **kwargs)
+        _set_span_errors(errors, span)
+        return errors
+
+
+def _traced_execute(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    if config.graphql.resolvers_enabled:
+        # patch resolvers
+        args, kwargs = _inject_trace_middleware_to_args(_resolver_middleware, args, kwargs)
+
+    # set resource name
+    if graphql_version < (3, 0):
+        document = get_argument_value(args, kwargs, 1, "document_ast")
+    else:
+        document = get_argument_value(args, kwargs, 1, "document")
+    resource = _get_source_str(document)
+
+    with pin.tracer.trace(
+        name="graphql.execute",
+        resource=resource,
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ) as span:
+        result = func(*args, **kwargs)
+        if isinstance(result, ExecutionResult):
+            # set error tags if the result contains a list of GraphqlErrors, skip if it's a promise
+            # TODO: support promises in graphql-core==2
+            _set_span_errors(result.errors, span)
+        return result
+
+
+def _traced_query(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    # set resource name
+    source = get_argument_value(args, kwargs, 1, "source")
+    resource = _get_source_str(source)
+
+    with pin.tracer.trace(
+        name="graphql.query",
+        resource=resource,
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ) as span:
+        _init_span(span)
+        result = func(*args, **kwargs)
+        if isinstance(result, ExecutionResult):
+            # set error tags if the result contains a list of GraphqlErrors, skip if it's a promise
+            _set_span_errors(result.errors, span)
```

Comment:
let's leave a comment about how we _could_ be duplicating error information from the child execute/validate/parse spans but we could also be catching other errors here.